cgen error when trying to `thread.wait()` on an unset thread field of struct

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.9 738f847, press to see full `v doctor` output</summary>

|V full version      |V 0.4.9 4225a34.738f847
|:-------------------|:-------------------
|OS                  |macos, macOS, 15.1.1, 24B91
|Processor           |8 cpus, 64bit, little endian, Apple M3
|Memory              |0.48GB/16GB
|                    |
|V executable        |/Users/adamoates/v/v
|V last modified time|2025-01-06 06:05:14
|                    |
|V home dir          |OK, value: /Users/adamoates/v
|VMODULES            |OK, value: /Users/adamoates/.vmodules
|VTMP                |OK, value: /tmp/v_501
|Current working dir |OK, value: /Users/adamoates/Documents/shattlebip
|                    |
|Git version         |git version 2.39.5 (Apple Git-154)
|V git status        |weekly.2024.53-37-g738f847f
|.git/config present |true
|                    |
|cc version          |Apple clang version 16.0.0 (clang-1600.0.26.4)
|gcc version         |Apple clang version 16.0.0 (clang-1600.0.26.4)
|clang version       |Apple clang version 16.0.0 (clang-1600.0.26.4)
|tcc version         |tcc version 0.9.28rc 2024-02-05 HEAD@105d70f7 (AArch64 Darwin)
|tcc git status      |thirdparty-macos-arm64 713692d4
|emcc version        |N/A
|glibc version       |N/A

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  /Users/adamoates/Documents/test.v && /Users/adamoates/Documents/test`
```v
struct Client {
mut:
	network_thread thread
}

fn main() {
	mut client := &Client{}
	client.network_thread.wait()
}

```

**What did you see?**
```
================== C compilation error (from cc): ==============
cc:       |                                                                                         ^~~
cc: /tmp/v_501/test.01JGX3FVGN7KNACR63BX3H1ZC6.tmp.c:5562:2: error: call to undeclared function '__v_thread_wait'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
cc:  5562 |         __v_thread_wait(client->network_thread);
cc:       |         ^
cc: 1 warning and 1 error generated.
... (the original output was 7 lines long, and was truncated to 5 lines)
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

Expected code to compile and when ran just skip right over `client.network_thead.wait()` since the thread has not been set. V would normally guard against this by not allowing you to set variables of uninitialized threads (i.e. no `mut t := thread{}`). It does not, however, check struct fields the same way. 


> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1581,6 +1581,9 @@ fn (mut g Gen) method_call(node ast.CallExpr) {
 			}
 		}
 		return
+	} else if typ_sym.info is ast.Thread {
+		waiter_fn_name := g.gen_gohandle_name(typ_sym.info.return_type)
+		g.create_waiter_handler(node.return_type, g.styp(typ_sym.info.return_type), waiter_fn_name)
 	}
 	left_sym := g.table.sym(left_type)
 	final_left_sym := g.table.final_sym(left_type)

--- a/vlib/v/gen/c/spawn_and_go.v
+++ b/vlib/v/gen/c/spawn_and_go.v
@@ -123,21 +123,7 @@ fn (mut g Gen) spawn_and_go_expr(node ast.SpawnExpr, mode SpawnGoMode) {
 	if g.pref.os == .windows && call_ret_type != ast.void_type {
 		g.writeln('${arg_tmp_var}->ret_ptr = (void *) _v_malloc(sizeof(${s_ret_typ}));')
 	}
-	is_opt := call_ret_type.has_flag(.option)
-	is_res := call_ret_type.has_flag(.result)
-	mut gohandle_name := ''
-	if call_ret_type == ast.void_type {
-		if is_opt {
-			gohandle_name = '__v_thread_Option_void'
-		} else if is_res {
-			gohandle_name = '__v_thread_Result_void'
-		} else {
-			gohandle_name = '__v_thread'
-		}
-	} else {
-		ret_styp := g.styp(g.unwrap_generic(call_ret_type)).replace('*', '_ptr')
-		gohandle_name = '__v_thread_${ret_styp}'
-	}
+	gohandle_name := g.gen_gohandle_name(call_ret_type)
 	if is_spawn {
 		if g.pref.os == .windows {
 			simple_handle := if node.is_expr && call_ret_type != ast.void_type {
@@ -182,48 +168,7 @@ fn (mut g Gen) spawn_and_go_expr(node ast.SpawnExpr, mode SpawnGoMode) {
 	g.writeln('// end go')
 	if node.is_expr {
 		handle = 'thread_${tmp}'
-		// create wait handler for this return type if none exists
-		waiter_fn_name := gohandle_name + '_wait'
-		mut should_register := false
-		lock g.waiter_fns {
-			if waiter_fn_name !in g.waiter_fns {
-				g.waiter_fns << waiter_fn_name
-				should_register = true
-			}
-		}
-		if should_register {
-			g.waiter_fn_definitions.writeln('${s_ret_typ} ${waiter_fn_name}(${gohandle_name} thread);')
-			g.gowrappers.writeln('\n${s_ret_typ} ${waiter_fn_name}(${gohandle_name} thread) {')
-			mut c_ret_ptr_ptr := 'NULL'
-			if call_ret_type != ast.void_type {
-				g.gowrappers.writeln('\t${s_ret_typ}* ret_ptr;')
-				c_ret_ptr_ptr = '&ret_ptr'
-			}
-			if g.pref.os == .windows {
-				if call_ret_type == ast.void_type {
-					g.gowrappers.writeln('\tu32 stat = WaitForSingleObject(thread, INFINITE);')
-				} else {
-					g.gowrappers.writeln('\tu32 stat = WaitForSingleObject(thread.handle, INFINITE);')
-					g.gowrappers.writeln('\tret_ptr = thread.ret_ptr;')
-				}
-			} else {
-				g.gowrappers.writeln('\tint stat = pthread_join(thread, (void **)${c_ret_ptr_ptr});')
-			}
-			g.gowrappers.writeln('\tif (stat != 0) { _v_panic(_SLIT("unable to join thread")); }')
-			if g.pref.os == .windows {
-				if call_ret_type == ast.void_type {
-					g.gowrappers.writeln('\tCloseHandle(thread);')
-				} else {
-					g.gowrappers.writeln('\tCloseHandle(thread.handle);')
-				}
-			}
-			if call_ret_type != ast.void_type {
-				g.gowrappers.writeln('\t${s_ret_typ} ret = *ret_ptr;')
-				g.gowrappers.writeln('\t_v_free(ret_ptr);')
-				g.gowrappers.writeln('\treturn ret;')
-			}
-			g.gowrappers.writeln('}')
-		}
+		g.create_waiter_handler(call_ret_type, s_ret_typ, gohandle_name)
 	}
 	// Register the wrapper type and function
 	mut should_register := false
@@ -411,9 +356,6 @@ fn (mut g Gen) spawn_and_go_expr(node ast.SpawnExpr, mode SpawnGoMode) {
 	}
 }
 
-// fn (mut g Gen) go_expr(node ast.GoExpr) {
-//}
-
 // get current thread size, if fn hasn't defined return default
 @[inline]
 fn (mut g Gen) get_cur_thread_stack_size(name string) string {
@@ -428,4 +370,68 @@ fn (mut g Gen) get_cur_thread_stack_size(name string) string {
 		}
 	}
 	return '${g.pref.thread_stack_size}'
+}
+
+fn (mut g Gen) gen_gohandle_name(typ ast.Type) string {
+	mut gohandle_name := ''
+	if typ == ast.void_type {
+		if typ.has_flag(.option) {
+			gohandle_name = '__v_thread_Option_void'
+		} else if typ.has_flag(.result) {
+			gohandle_name = '__v_thread_Result_void'
+		} else {
+			gohandle_name = '__v_thread'
+		}
+	} else {
+		ret_styp := g.styp(g.unwrap_generic(typ)).replace('*', '_ptr')
+		gohandle_name = '__v_thread_${ret_styp}'
+	}
+	return gohandle_name
+}
+
+fn (mut g Gen) create_waiter_handler(call_ret_type ast.Type, s_ret_typ string, gohandle_name string) {
+	// create wait handler for this return type if none exists
+	waiter_fn_name := gohandle_name + '_wait'
+	mut should_register := false
+	lock g.waiter_fns {
+		if waiter_fn_name !in g.waiter_fns {
+			g.waiter_fns << waiter_fn_name
+			should_register = true
+		}
+	}
+	if !should_register {
+		return
+	}
+	g.waiter_fn_definitions.writeln('${s_ret_typ} ${waiter_fn_name}(${gohandle_name} thread);')
+	g.gowrappers.writeln('\n${s_ret_typ} ${waiter_fn_name}(${gohandle_name} thread) {')
+	mut c_ret_ptr_ptr := 'NULL'
+	if call_ret_type != ast.void_type {
+		g.gowrappers.writeln('\t${s_ret_typ}* ret_ptr;')
+		c_ret_ptr_ptr = '&ret_ptr'
+	}
+	if g.pref.os == .windows {
+		if call_ret_type == ast.void_type {
+			g.gowrappers.writeln('\tu32 stat = WaitForSingleObject(thread, INFINITE);')
+		} else {
+			g.gowrappers.writeln('\tu32 stat = WaitForSingleObject(thread.handle, INFINITE);')
+			g.gowrappers.writeln('\tret_ptr = thread.ret_ptr;')
+		}
+	} else {
+		g.gowrappers.writeln('\tif ((unsigned long int)thread == 0) { _v_panic(_SLIT("unable to join thread")); }')
+		g.gowrappers.writeln('\tint stat = pthread_join(thread, (void **)${c_ret_ptr_ptr});')
+	}
+	g.gowrappers.writeln('\tif (stat != 0) { _v_panic(_SLIT("unable to join thread")); }')
+	if g.pref.os == .windows {
+		if call_ret_type == ast.void_type {
+			g.gowrappers.writeln('\tCloseHandle(thread);')
+		} else {
+			g.gowrappers.writeln('\tCloseHandle(thread.handle);')
+		}
+	}
+	if call_ret_type != ast.void_type {
+		g.gowrappers.writeln('\t${s_ret_typ} ret = *ret_ptr;')
+		g.gowrappers.writeln('\t_v_free(ret_ptr);')
+		g.gowrappers.writeln('\treturn ret;')
+	}
+	g.gowrappers.writeln('}')
 }

--- a/vlib/v/slow_tests/inout/thread_uninitialized_err.out
+++ b/vlib/v/slow_tests/inout/thread_uninitialized_err.out
@@ -1,0 +1,2 @@
+================ V panic ================
+V panic: unable to join thread

--- a/vlib/v/slow_tests/inout/thread_uninitialized_err.vv
+++ b/vlib/v/slow_tests/inout/thread_uninitialized_err.vv
@@ -1,0 +1,10 @@
+struct Client {
+mut:
+	network_thread thread
+}
+
+fn main() {
+	mut client := &Client{}
+	println('================ V panic ================')
+	client.network_thread.wait()
+}

